inlanecruising_plugin/src/inlanecruising_plugin.cpp

```diff
@@ -18,253 +18,507 @@
+  }
+  return std::sqrt(fabs(lateral_accel_limit / curvature));
+}
+std::vector<double> compute_ideal_speeds(std::vector<double> curvatures, double lateral_accel_limit)
+{
+  std::vector<double> out;
+  for (double k : curvatures)
+  {
+    out.push_back(compute_speed_for_curvature(k, lateral_accel_limit));
+  }
 
-    void InLaneCruisingPlugin::initialize()
+  return out;
+}
+
+std::vector<double> apply_speed_limits(const std::vector<double> speeds, const std::vector<double> speed_limits)
+{
+  ROS_ERROR_STREAM("Speeds list size: " << speeds.size());
+  ROS_ERROR_STREAM("SpeedLimits list size: " << speed_limits.size());
+  if (speeds.size() != speed_limits.size())
+  {
+    throw std::invalid_argument("Speeds and speed limit lists not same size");
+  }
+  std::vector<double> out;
+  for (int i = 0; i < speeds.size(); i++)
+  {
+    out.push_back(std::min(speeds[i], speed_limits[i]));
+  }
+
+  return out;
+}
+
+std::vector<double> compute_downtracks(std::vector<lanelet::BasicPoint2d> basic_points)
+{
+  std::vector<double> downtracks;
+  downtracks.reserve(basic_points.size());
+  double current_dt = 0;
+  boost::optional<lanelet::BasicPoint2d> prev_p;
+  for (const auto& p : basic_points)
+  {
+    if (!prev_p)
     {
-        nh_.reset(new ros::CARMANodeHandle());
-        pnh_.reset(new ros::CARMANodeHandle("~"));
-        
-        trajectory_srv_ = nh_->advertiseService("plugins/InLaneCruisingPlugin/plan_trajectory", &InLaneCruisingPlugin::plan_trajectory_cb, this);
-                
-        inlanecruising_plugin_discovery_pub_ = nh_->advertise<cav_msgs::Plugin>("plugin_discovery", 1);
-        plugin_discovery_msg_.name = "InLaneCruisingPlugin";
-        plugin_discovery_msg_.versionId = "v1.0";
-        plugin_discovery_msg_.available = true;
-        plugin_discovery_msg_.activated = false;
-        plugin_discovery_msg_.type = cav_msgs::Plugin::TACTICAL;
-        plugin_discovery_msg_.capability = "tactical_plan/plan_trajectory";
-
-        waypoints_sub_ = nh_->subscribe("final_waypoints", 1, &InLaneCruisingPlugin::waypoints_cb, this);
-        pose_sub_ = nh_->subscribe("current_pose", 1, &InLaneCruisingPlugin::pose_cb, this);
-        twist_sub_ = nh_->subscribe("current_velocity", 1, &InLaneCruisingPlugin::twist_cd, this);
-        pnh_->param<double>("trajectory_time_length", trajectory_time_length_, 6.0);
-        pnh_->param<double>("trajectory_point_spacing", trajectory_point_spacing_, 0.1);
-
-        ros::CARMANodeHandle::setSpinCallback([this]() -> bool {
-            inlanecruising_plugin_discovery_pub_.publish(plugin_discovery_msg_);
-            return true;
-        });
+      downtracks.push_back(0);
+      continue;
     }
+    double dx = p.x() - prev_p->x();
+    double dy = p.y() - prev_p->y();
+    double dist = sqrt(dx * dx + dy * dy);
+    current_dt += dist;
+    downtracks.push_back(current_dt);
+    prev_p = p;
+  }
+
+  return downtracks;
+}
 
+struct DiscreteCurve
```

Comment:
Not sure I love the name, as since this contains speeds as well it's a bit more than just a curve. But it's fine for now, that's a minor quibble.

Reply:
Easy enough to change if you have another name in mind